Add the process of granting API Pull access to the onboarding flow

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Project thread: pcTzPl-2Df-p2

The PR adds the process of granting API Pull access to step 1 of the onboarding flow.

### Screenshots:

#### 📹 Approve Google account and WPCOM app authorizations with all required access

This demonstrates the case where the user is automatically and seamlessly redirected to WPCOM app authorization once they are being redirected back from the Google authorization with all required access.

https://github.com/user-attachments/assets/b6d101d8-6387-4446-9df2-f3f154a025c5

#### 📹 Approve Google account authorization with partial access, and deny WPCOM app authorization on the first attempt

This demonstrates the case:

- The mid-state asks the user to grant all required access to their Google account
- The mid-state asks the user again to approve WPCOM app authorization

https://github.com/user-attachments/assets/7d589351-7b15-4f12-b3d3-60635dfd4b3f

#### 📷 The default mid-state UI without additional messages

User will see it in the following scenarios:

- User interrupted the onboarding process after completing the Google account connection and authorization, and then return to the onboarding process again.
- In any case the authorization URL in the automated and seamless process was not successfully obtained.

![image](https://github.com/user-attachments/assets/e671dc52-4efc-445a-ae0b-29524f29515a)

#### 📷 The mid-state UI after the user denied authorization

![image](https://github.com/user-attachments/assets/d52ee42f-e81e-4c10-9717-ff7f37a9b65c)

#### 📷 The mid-state UI after the user approved authorization with a mismatched WPCOM account

The user has connected a WPCOM account *A* before, and they approved authorization with another WPCOM account *B*

![image](https://github.com/user-attachments/assets/d39a8bd5-9c73-4afc-a08f-d58ebf738514)

### Detailed test instructions:

1. Use the Connection Test page to disconnect Google account and the authorization of Google's WPCOM app if they have been connected or granted
2.Make the test site publicly accessible
3. Go to step 1 of the onboarding flow
4. Connect to Google account with the all required permissions, and check if it automatically redirects you to WPCOM app authorization once you are being redirected back from the Google authorization
5. Approve WPCOM app authorization
6. After back to the onboarding flow, check if the connection setup of Google Merchant Center and Google Ads accounts are shown
7. Click the "Or, connect to a different Google account" button to and redo step 4
   - This step and the next step will verify if the WPCOM app is also disconnected as well
8. Deny WPCOM app authorization
9. After back to the onboarding flow, check if the Google account card is shown with the mid-state UI showing
   > Access was denied. Please make sure to grant Google access to your WooCommerce store to continue.
10. Click the "Grant access" button to see if it redirects you to the authorization of WPCOM app
11. Approve WPCOM app authorization
12. After back to the onboarding flow, check if the connection setup of Google Merchant Center and Google Ads accounts are shown
13. Complete the rest account connections and check to see if the "Continue" button at the bottom is enabled
14. Continue to step 2
15. Refresh the webpage to check if it starts from step 2
16. View the E2E testing result: https://github.com/woocommerce/google-listings-and-ads/actions/runs/13564482640/job/37914394871

### Changelog entry

> Add - During onboarding, it requires granting access to Google's WordPress.com application to synchronize product data with Google.

eason9487 committed:
Make the `step` property of `mc/setup` API also determine the completion of API Pull authorization.

The `step` value should be 'accounts' when the API Pull is not yet authorized.

eason9487 committed:
Add a new property `isWPComAppGranted` to the return payload of `useGoogleMCAccount`.

eason9487 committed:
Implement a new component `WPComAppAuthorizationCard`.

eason9487 committed:
Add the process of granting API Pull access to the onboarding flow.

eason9487 committed:
Update event tracking.

eason9487 committed:
Update E2E tests.

eason9487 committed:
In onboarding flow, add a process to seamlessly and automatically redirect to WPCOM app authorization.

eason9487 committed:
Add status notices to the `AuthorizeWPComAppCard` component.

eason9487 committed:
Move the `authorize-wpcom-app-card.js` file to the `google-account-card` directory.

eason9487 committed:
Finalize the appearance and wording of the `AuthorizeWPComAppCard` component.

eason9487 committed:
Disconnect WPCOM app as well when disconnecting Google account during onboarding.

ianlin commented on js/src/pages/onboarding/index.js:
```diff
@@ -12,6 +14,15 @@ import SetupStepper from './setup-stepper';
+	// Render nothing as the requirement is to make it look like the redirections
+	// between Google authorization and WPCOM app authorization are seamless.
+	if ( ! canContinueRendering ) {
+		return null;
```
Not sure if we could add a spinner here since the blank page before redirecting to WPCOM feels a bit long in my site.

eason9487 replied:
Thanks for the suggestion!

I also needed to wait for a while during this process. For a moment I would suspect the blank page was stuck or not responding. Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/2802/commits/77e5beb6ae5d951fb384f5e0b666415bbd551fa0

https://github.com/user-attachments/assets/cc43fc15-7119-4f14-b01b-1b3582735947

ianlin reviewed (approved):
Thanks for adding the granting process in onboarding flow, it looks nice. I tested all the scenarios and they all worked correctly. Left a non-blocking comment and approving it.

eason9487 committed:
Render a spinner instead of nothing when waiting for the result of `useAutoWPComAppAuthorization`.

Address: https://github.com/woocommerce/google-listings-and-ads/pull/2802#discussion_r1978580452
